src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,93 @@
+
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions.LanguageServices
+{
+    /// <summary>
+    /// Analyzer that reports diagnostics in strings that we know are regex text.
+    /// </summary>
+    internal class RegexDiagnosticAnalyzer : IEmbeddedDiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "RE001";
+
+        private readonly RegexEmbeddedLanguage _language;
+        private readonly DiagnosticDescriptor _descriptor;
+
+        public RegexDiagnosticAnalyzer(RegexEmbeddedLanguage language)
+        {
+            _language = language;
+
+            _descriptor = new DiagnosticDescriptor(DiagnosticId,
+                new LocalizableResourceString(nameof(WorkspacesResources.Regex_issue_0), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)),
+                new LocalizableResourceString(nameof(WorkspacesResources.Regex_issue_0), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)),
+                "REGEX",
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault: true);
+
+            SupportedDiagnostics = ImmutableArray.Create(_descriptor);
+        }
+
+        public ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public void Analyze(SemanticModelAnalysisContext context, OptionSet optionSet)
+        {
+            var semanticModel = context.SemanticModel;
+            var syntaxTree = semanticModel.SyntaxTree;
+            var cancellationToken = context.CancellationToken;
+            var options = context.Options;
+
+            var option = optionSet.GetOption(RegularExpressionsOptions.ReportInvalidRegexPatterns, syntaxTree.Options.Language);
+            if (!option)
+            {
+                return;
+            }
+
+            var detector = RegexPatternDetector.TryGetOrCreate(semanticModel, _language);
+            if (detector == null)
+            {
+                return;
+            }
+
+            var root = syntaxTree.GetRoot(cancellationToken);
+            Analyze(context, detector, root, cancellationToken);
+        }
+
+        private void Analyze(
+            SemanticModelAnalysisContext context, RegexPatternDetector detector,
+            SyntaxNode node, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            foreach (var child in node.ChildNodesAndTokens())
+            {
+                if (child.IsNode)
+                {
+                    Analyze(context, detector, child.AsNode(), cancellationToken);
+                }
+                else
+                {
+                    var token = child.AsToken();
+                    if (token.RawKind == _language.StringLiteralKind)
```

Comment:
>StringLiteralKind [](start = 51, length = 17)

I thought analyzers have some infrastructure to declare which nodes they are interested in. Why can't we leverage that here? Is it a token vs. node issue?

Reply:
The issue here is perf.  We have something expensive we want to do once per SemanticModel (i.e. figuring out the acutal System regex types and whatnot).  And then we want to blast through and check all the strings. I don't want to perform all that common work over and over and over again on each string.  

Similarly, i don't want to have to go check the document+options on each and every string token i run into to determine if the feature is running.  

So, instead, i just made this a document-oriented analyzer.  I can do all the expensive work once, then blast through and check each token cheaply.